File: mine/sorts/minheap.py
# element in index 0 is not the heap element
class MinHeap:
    
    def __init__(self, capacity):
        self.capacity = capacity
        self.heap = []

    def Dump(self):
        print(self.heap[1:])

    def Min(self):
        return self.heap[1]

    def Swim(self, i):
        t = self.heap[i]
        while i > 1 and self.heap[i // 2] > self.heap[i]:
            self.heap[i] = self.heap[i // 2]
            i = i // 2
        self.heap[i] = t

    def Sink(self, i):
        t = self.heap[i]
        while 2 * i <= len(self.heap) - 1:  # node i has leaf
            child = 2 * i
            if child < len(self.heap) - 1 and self.heap[child + 1] < self.heap[child]:
                child = child + 1
            if self.heap[child] < t:
                self.heap[i] = self.heap[child]
                i = child
            else:
                break;
        self.heap[i] = t
    
    def Min(self):
        return self.heap[1]

    def Push(self, t):
        self.heap.append(t)
        self.Swim(len(self.heap) - 1)

    def Pop(self):
        self.heap[1] = self.heap[len(self.heap) - 1]
        self.heap.pop()
        self.Sink(1)

    def BuildHeap(self, arr):
        self.heap.append(-1)
        for i in range(len(arr)):
            self.heap.append(arr[i])

        for i in range(len(arr) // 2, 0, -1):
            self.Sink(i)


if __name__ == '__main__':
    arr = [-100, 3, 88, 67, -8, 0, 99, 299, 9, -111]
    heap = MinHeap(20)
    heap.BuildHeap(arr)
    heap.Dump()

File: mine/sorts/minheap.rs
#![allow(dead_code)]
use std::fmt::Display;

struct MinHeap<T> {
    heap: Vec<T>,
}

impl<T: Copy + PartialOrd + Display> MinHeap<T> {
    fn new() -> MinHeap<T> {
        MinHeap {
            heap: Vec::new(),
        }
    }

    fn build(&mut self, arr: &[T]) {
        self.heap.push(arr[0]);
        for i in arr.iter() {
            self.heap.push(*i);
        }

        let ib = arr.len() / 2 + 1;
        for i in (1..ib).rev() {
            self.sink(i);
        }
    }

    fn new2(capacity: usize) -> MinHeap<T> {
        MinHeap {
            heap: Vec::with_capacity(capacity),
        }
    }

    fn min(&self) -> &T {
        &self.heap[1]
    }

    fn push(&mut self, t: &T) {
        self.heap.push(*t);
        let i = self.heap.len() - 1;
        self.swim(i);
        //self.swim(self.heap.len() - 1); //error self.heap.len() --> immutable borrow
    }

    fn pop(&mut self) {
        if let Some(last) = self.heap.pop() {
            self.heap[1] = last;
            self.sink(1);
        }
    }

    fn sink(&mut self, mut i: usize) {
        let t: T = self.heap[i].clone();
        while 2 * i <= self.heap.len() - 1 { // node i have leaf
            let mut child = 2 * i;
            // node i have right leaf and right leaf is less than left leaf
            if child < self.heap.len() - 1 && self.heap[child] > self.heap[child + 1] { 
                child = child + 1;
            }

            if &self.heap[i] > &self.heap[child] {
                self.heap[i] = self.heap[child].clone();
                i = child;
            } else {
                break;
            }
        }
        self.heap[i] = t;
    }

    fn swim(&mut self, mut i: usize) {
        let t: T = self.heap[i].clone();
        while i > 1 && &self.heap[i / 2] < &self.heap[i] {
            self.heap[i] = self.heap[i / 2].clone();
            i = i / 2;
        }
        self.heap[i] = t;
    }

    fn dump(&self) {
        for i in 1..self.heap.len() {
            println!("{} ", self.heap[i]);
        }
    }

    fn move_to(&mut self, other: Vec<T>) {
        self.heap = other;
    }
}

fn main() {
    let mut heap1: MinHeap<i32> = MinHeap::new();
    heap1.build(&vec![-11, 3, 0, 88, -87, 1]);
    heap1.dump();

    println!("ttttttttttttttttttttttt");

    let mut heap2: MinHeap<i32> = MinHeap::new();
    heap2.build(&[19, -11, 3, 0, 88, -87, 1]);
    heap2.dump();
}
